File: src/main/python/basic/segment_tree.py
import random
from functools import reduce


class SegmentTree:
    """
    This is an example of SUM Tree using segment tree.
    """
    def __init__(self):
        self.value = 0
        self.left = None
        self.right = None

    def __str__(self):
        return " %d " % self.value


def fill_items(root, array, l, r):
    if l == r:
        root = SegmentTree()
        root.value = array[l]
        return root
    else:
        mid = (l + r) / 2
        if root is None:
            root = SegmentTree()
        root.left = fill_items(root.left, array, l, mid)
        root.right = fill_items(root.right, array, mid + 1, r)
        root.value = root.left.value + root.right.value
        return root


def build_segment_tree(array):
    return fill_items(None, array, 0, len(array) - 1)


def get_sum(node, ci, cj, i, j):
    if ci < i and cj < i:
        return 0
    if ci > j and cj > j:
        return 0
    if i <= ci and cj <= j:
        return node.value
    mid = (ci + cj) / 2
    p1 = get_sum(node.left, ci, mid, i, j)
    p2 = get_sum(node.right, mid + 1, cj, i, j)
    return p1 + p2


def update(root, l, r, index_to_update, increase_by):
    if root is None:
        return
    if l == r:
        root.value = root.value + increase_by
        return

    mid = (l + r) / 2
    if l <= index_to_update <= mid:
        update(root.left, l, mid, index_to_update, increase_by)
        root.value = root.left.value + root.right.value
    elif mid < index_to_update <= r:
        update(root.right, mid + 1, r, index_to_update, increase_by)
        root.value = root.left.value + root.right.value


if __name__ == '__main__':

    array = map(lambda x: random.randint(1, 100), range(100))
    size = len(array)
    ci, cj = 0, size - 1
    i, j = 1, 88
    tree = build_segment_tree(array)
    print (reduce(lambda x, y: x + y, array[i:j + 1]))
    print (get_sum(tree, ci, cj, i, j))

    index_to_update = 50
    increase_value_by = 17

    array[index_to_update] = array[index_to_update] + increase_value_by
    update(tree, ci, cj, index_to_update, increase_value_by)

    print (reduce(lambda x, y: x + y, array[i:j + 1]))
    print (get_sum(tree, ci, cj, i, j))

File: src/main/java/com/personal/SegmentTree.java
package com.personal;

import java.util.HashMap;
import java.util.Map;


public class SegmentTree {

    private static int N = 10;
    Map<Integer, Integer> tree;
    Map<Integer, Integer> lazy;

    public SegmentTree() {
        tree = new HashMap<>();
        lazy = new HashMap<>();
    }

    void push(int i, int l, int r) {

        Integer value = lazy.get(i);
        if (value == null) {
            return;
        }

        // update current node
        // update will be from all the nodes within the range. hence multiply
        // in case of Segment Sum Tree
        tree.put(i, tree.getOrDefault(i, 0) + (r - l + 1) * value);

        if (l != r) {
            int left = 2 * i + 1;
            int right = 2 * i + 2;

            lazy.put(left, lazy.getOrDefault(left, 0) + value);
            lazy.put(right, lazy.getOrDefault(right, 0) + value);
        }
        // clear the current lazy node
        lazy.remove(i);
    }

    public void updateRange(int ql, int qr, int value) {
        updateRange(0, 0, N, ql, qr, value);
    }

    private void updateRange(int idx, int l, int r, int ql, int qr, int value) {
        push(idx, l, r);

        if (qr < l || r < ql) {
            return;
        }

        if (ql <= l && r <= qr) {
            lazy.put(idx, lazy.getOrDefault(idx, 0) + value);
            push(idx, l, r);
            return;
        }

        int mid = l + (r - l) / 2;
        updateRange(2 * idx + 1, l, mid, ql, qr, value);
        updateRange(2 * idx + 2, mid + 1, r, ql, qr, value);

        int lval = tree.getOrDefault(2 * idx + 1, 0);
        int rval = tree.getOrDefault(2 * idx + 2, 0);

        tree.put(idx, lval + rval);
    }

    public int queryRange(int l, int r) {
        return queryRange(0, 0, N, l, r);
    }

    private int queryRange(int idx, int l, int r, int ql, int qr) {
        push(idx, l, r);

        if (qr < l || r < ql) {
            return 0;
        }

        if (ql <= l && r <= qr) {
            return tree.getOrDefault(idx, 0);
        }

        int mid = l + (r - l) / 2;
        return queryRange(2 * idx + 1, l, mid, ql, qr)
                + queryRange(2 * idx + 2, mid + 1, r, ql, qr);
    }

    public static void main(String[] args) {
        SegmentTree seg = new SegmentTree();

        seg.updateRange(2, 4, 5);

        // Add 3 to [3,10]
        seg.updateRange(3, 10, 3);

        // Queries
        System.out.println(seg.queryRange(2, 4));
        System.out.println(seg.queryRange(3, 3));
        System.out.println(seg.queryRange(0, 10));
    }

}
